aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu

```diff
@@ -0,0 +1,326 @@
+        const int64_t *partials_per_segment_offset,
+        const int64_t *segment_offsets,
+        int64_t num_of_segments) {
+  const int id = blockIdx.x * blockDim.x + threadIdx.x;
+  if(id < num_of_segments) {
+    int64_t idx = partials_per_segment_offset[id];
+    const int64_t num_partials = partials_per_segment[id];
+    const int64_t segment_offset = segment_offsets[id];
+    for (int64_t i=0; i<num_partials; ++i) {
+      ret[idx++] = segment_offset + i * NROWS_PER_THREAD;
+    }
+  }
+}
+
+
+template <typename scalar_t>
+__global__ void compute_grad_weight_bags(
+    int64_t *indices, scalar_t *gradOutput,
+    int64_t *offset2bag, int64_t *count, ptrdiff_t numel,
+    int64_t stride, int mode_mean, const int64_t *bag_size,
+    scalar_t* per_sample_weights, int64_t per_sample_weights_stride,
+    int64_t* segment_offsets, int64_t num_of_segments, scalar_t *grad_weight_per_segment,
+    const int64_t stride_warped) {
+
+  const int gid = blockIdx.x * blockDim.x + threadIdx.x;
+  const int id = gid / stride_warped;
+  const int startFeature = gid % stride_warped;
+  if (startFeature >= stride) {
+    return;
+  }
+  if (id >= num_of_segments) {
+    return;
+  }
+  const int idx_begin = segment_offsets[id];
+  const int idx_end = (id == num_of_segments-1)?numel:segment_offsets[id+1];
+
+  acc_type<scalar_t, true> weight = 0;
+  for (int idx=idx_begin; idx < idx_end; ++idx) {
+    const int origRow = indices[idx];
+    const int seq_number = offset2bag[origRow];
+    const int gradOutputRow = seq_number * stride;
+
+    acc_type<scalar_t, true> scale = count ? 1.0 / count[idx] : 1.0;
+    if (per_sample_weights) {
+      scale *= per_sample_weights[origRow * per_sample_weights_stride];
+    }
+
+    acc_type<scalar_t, true> gradient = gradOutput[gradOutputRow + startFeature];
+    if (mode_mean) {
+      gradient /= bag_size[seq_number];
+    }
+    weight += gradient * scale;
+  }
+  grad_weight_per_segment[id * stride + startFeature] = weight;
+}
+
+template <typename scalar_t>
+__global__ void compute_grad_weight(
+    int64_t *indices,
+    scalar_t *gradOutput,
+    int64_t *count,
+    ptrdiff_t numel,
+    int64_t stride,
+    int64_t* segment_offsets,
+    int64_t num_of_segments,
+    scalar_t *grad_weight_per_segment,
+    int padding_idx,
+    const int64_t stride_warped) {
+
+  using accscalar_t = acc_type<scalar_t, true>;
+  const int gid = blockIdx.x * blockDim.x + threadIdx.x;
+  const int id = gid / stride_warped;
+  const int startFeature = gid % stride_warped;
+  if (startFeature >= stride) {
+    return;
+  }
+  if (id >= num_of_segments) {
+    return;
+  }
+  const int idx_begin = segment_offsets[id];
+  const int idx_end = (id == num_of_segments-1)?numel:segment_offsets[id+1];
+  if (idx_begin == padding_idx) {
+    return;
+  }
+
+  accscalar_t weight = 0;
+  for (int idx=idx_begin; idx < idx_end; ++idx) {
+    const int64_t target_row = indices[idx];
+    if (target_row != padding_idx) {
```

Comment:
target_row is the same for all indices in the loop, thus this check should be moved to before the loop. I agree it does not affect correctness and performance, but it affects code readability.

Reply:
Fixed in #22401